? sites/default/files
? sites/default/settings.php
Index: modules/taxonomy/taxonomy.module
===================================================================
RCS file: /cvs/drupal/drupal/modules/taxonomy/taxonomy.module,v
retrieving revision 1.597
diff -u -p -r1.597 taxonomy.module
--- modules/taxonomy/taxonomy.module	27 Jun 2010 16:49:58 -0000	1.597
+++ modules/taxonomy/taxonomy.module	29 Jun 2010 11:23:40 -0000
@@ -738,11 +738,18 @@ function taxonomy_get_parents_all($tid) 
 
   $parents = array();
   if ($term = taxonomy_term_load($tid)) {
-    $parents[] = $term;
-    $n = 0;
-    while ($parent = taxonomy_get_parents($parents[$n]->tid)) {
-      $parents = array_merge($parents, $parent);
-      $n++;
+    $parents[$term->tid] = $term;
+    $terms_to_search[] = $term->tid;
+
+    while ($tid = array_shift($terms_to_search)) {
+      if ($new_parents = taxonomy_get_parents($tid)) {
+        foreach ($new_parents as $new_parent) {
+          if (! isset($parents[$new_parent->tid])) {
+            $parents[$new_parent->tid] = $new_parent;
+            $terms_to_search[] = $new_parent->tid;
+          }
+        }
+      }
     }
   }
 
Index: modules/taxonomy/taxonomy.test
===================================================================
RCS file: /cvs/drupal/drupal/modules/taxonomy/taxonomy.test,v
retrieving revision 1.80
diff -u -p -r1.80 taxonomy.test
--- modules/taxonomy/taxonomy.test	28 May 2010 10:17:02 -0000	1.80
+++ modules/taxonomy/taxonomy.test	29 Jun 2010 11:23:40 -0000
@@ -420,6 +420,13 @@ class TaxonomyTermTestCase extends Taxon
     taxonomy_term_save($term2);
     $parents = taxonomy_get_parents($term2->tid);
     $this->assertTrue(isset($parents[$term1->tid]) && isset($parents[$term3->tid]), t('Both parents found successfully.'));
+    // Do the same check with taxonomy_get_parents_all().
+    $parents = taxonomy_get_parents_all($term2->tid);
+    $this->assertTrue(isset($parents[$term1->tid]) && isset($parents[$term3->tid]), t('Both parents found successfully.'));
+
+    // Try to get parents of an invalid term.
+    $this->assertFalse(taxonomy_get_parents(6000), t('Parents were not loaded for an invalid term ID.'));
+    $this->assertFalse(taxonomy_get_parents_all(6000), t('Parents were not loaded for an invalid term ID.'));
   }
 
   /**
